[terryasdf] iP

terryasdf opened the pull request:
Finish week 3

terryasdf committed:
Initialize Chatbot Terry

rename all 'Duke' related symbols to 'Terry'
implement initial skeletal version chatbot

terryasdf committed:
Create functions for print greeting and exit messages.

terryasdf committed:
Add readCmd skeleton

terryasdf committed:
Implement features: add, list

add addToCmdList and printCmdList

terryasdf committed:
Task.java: Create Task class replacing cmd

terryasdf committed:
Terry.java: Implement feature: mark/unmark tasks

change datatype of cmdList to Task
rename functions
add function: getFormattedTaskInfo for printing tasks
modify list command to show isDone status
add functions: markTask, unmarkTask

terryasdf committed:
Terry.java: Refactor code

create constants for output messages
make sure constants' names comply with Java standards

Hudou0420 commented on src/main/java/Terry.java:
```diff
@@ -0,0 +1,94 @@
+    private final static String GREETING_MSG = "Hello! I'm Terry!\nWhat can I do for you?";
+    private final static String EXIT_MSG = "Bye. Hope to see you again soon!";
+    public static final String MARK_TASK_MSG = "Task marked as done. Nice job.";
+    public static final String UNMARK_TASK_MSG = "Task marked as not done.";
```
Good use of static final to use later in the code!

Hudou0420 commented on src/main/java/Terry.java:
```diff
@@ -0,0 +1,94 @@
+
+    private static String getFormattedTaskInfo(int id) {
+        if (id < 0 || id >= taskCount) return "";
+        return "[" + (tasks[id].isDone() ? 'X' : ' ') + "] " + tasks[id].getDescription();
```
suggestion
try to separate this line to avoid long and complicated line. It might take more time for others to read and understand what it is doing exactly.

Hudou0420 commented on src/main/java/Terry.java:
```diff
@@ -0,0 +1,94 @@
+        Scanner in = new Scanner(System.in);
+
+        while (true) {
+            String cmd = in.nextLine();
```
suggestion
maybe rename this variable to cmdInput to make it clear that it is an input from the user(?)

Hudou0420 commented on src/main/java/Terry.java:
```diff
@@ -0,0 +1,94 @@
+            if (cmdArgs.length == 0) continue;
+
+            System.out.println(SPLIT_LINE);
+            if (cmd.equals("bye")) {
+                printExitMsg();
+                in.close();
+                return;
+            } else if (cmd.equals("list")) {
+                printTasks();
+            } else if (cmdArgs[0].equals("mark")) {
+                if (cmdArgs.length < 2) continue;
+                int id = Integer.parseInt(cmdArgs[1]);
+                markTask(id);
+            } else if (cmdArgs[0].equals("unmark")) {
+                if (cmdArgs.length < 2) continue;
+                int id = Integer.parseInt(cmdArgs[1]);
+                unmarkTask(id);
+            } else {
+                addTask(cmd);
+            }
+            System.out.println(SPLIT_LINE);
```
suggestion
can try to avoid so many if else by usinng switch statement if possible

Hudou0420 reviewed (commented):
Looks good to me

terryasdf committed:
Create classes: Deadline, Event from ToDo

rename Task to ToDo

terryasdf committed:
Refactor code

create Cmd for commands
create Msg for output messages
create TerryController for data access
create TerryService for handling data and message generation
create MsgHandler for handling messages
create TerryFramework for managing CLI

terryasdf committed:
Implement ToDos, Events and Deadlines support

add commands: todo, ddl, event
CmdParser.java: revise argument parsing

terryasdf committed:
Change message format for command: list

terryasdf committed:
Revise cmd parsing logic

terryasdf committed:
Implement scripted test

runtest.sh: add -Z option to diff for ignoring trailing whitespace
MsgString.java: replace all UTF-8 characters

terryasdf commented:
Finish week 4

swethacool commented on src/main/java/Terry.java:
```diff
@@ -2,22 +2,27 @@
+    public static final String UNMARK_TASK_MSG = "Task marked as not done.";
+
+    private final static int MAX_TASKS = 100;
+    private final static Task[] tasks = new Task[MAX_TASKS];
```
Good practice to use a constant (MAX_TASKS) instead of hardcoding 100 multiple times.

swethacool commented on src/main/java/Terry.java:
```diff
@@ -2,22 +2,27 @@
+        System.out.println(SPLIT_LINE);
+        System.out.println(EXIT_MSG);
+        System.out.println(SPLIT_LINE);
     }
```
Extracting greeting and exit messages into separate methods improves modularity and avoids duplication.

swethacool commented on src/main/java/Terry.java:
```diff
@@ -41,14 +46,14 @@ private static void addTask(String description) {
     private static void markTask(int id) {
         if (id < 0 || id >= taskCount) return;
```
Good validation to prevent out-of-bounds access to the tasks array.

swethacool replied:
Suggestion: to print out an error message if an invalid id is given. Otherwise the method silently fails without informing the user.

swethacool commented on src/main/java/Terry.java:
```diff
@@ -2,22 +2,27 @@
 
-    private final static String splitLine = "____________________________________________________________";
-    private final static Task[] tasks = new Task[100];
+    private final static String SPLIT_LINE = "____________________________________________________________";
```
splitLine and SPLIT_LINE are duplicates. For example its better to remove splitLine and use SPLIT_LINE consistently.

swethacool commented on src/main/java/Terry.java:
```diff
@@ -78,7 +83,7 @@ private static void readCmd() {
                 addTask(cmd);
             }
-            System.out.println(splitLine);
+            System.out.println(SPLIT_LINE);
```
Suggestion to use a switch statement for better readability.

swethacool reviewed (commented):
Overall I found the code easy to read and manageable due to breaking it down into smaller methods so as to prevent duplication as well.

swethacool commented on src/main/java/Terry.java:
```diff
@@ -78,7 +83,7 @@ private static void readCmd() {
+            System.out.println(SPLIT_LINE);
         }
     }
 
```
Suggestion to use switch statements as too many if-else statements hinder the ability to read the code.

terryasdf committed:
Refactor code: package classes

terryasdf committed:
terry.framework: Fix wrong class name for Controller and Service

terryasdf committed:
Add comments to classes and methods

terryasdf committed:
terry.framework.Service: Change return type of methods

runtest.bat: update compile path

terryasdf committed:
terry.exception: Create exception class: UnknownCmdKeywordException

terry.exception: create class: ExceptionHandler
terry.msg.ReturnStatus: add corresponding return status

terryasdf committed:
Implement multiple exceptions

create class: ExceptionHandler
create class: CmdRunner
terry.cmd.CmdParser: update argument parsing logic

terryasdf committed:
Merge branch 'branch-Level-5'

terryasdf committed:
Fix outputs

terryasdf commented:
Finish week 5

terryasdf committed:
terry.framework: Implement command: delete

terry.framework.Controller: replace array with ArrayList

terryasdf committed:
Implement command: save csv

create class: FileHandler
terry.exception.ExceptionHandler: add method to handle IOException

terryasdf committed:
Merge branch 'branch-Level-6'

terryasdf committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/terry/cmd/CmdKeyword.java
#	src/main/java/terry/framework/CmdRunner.java
#	src/main/java/terry/msg/MsgString.java

terryasdf committed:
Add manifest info for jar release

terryasdf committed:
Add lib: org.json

terryasdf committed:
Replace csv with json

terryasdf committed:
Implement command: load

refactor terry.framework.Framework and terry.framework.Runtime

terryasdf committed:
Implement command: find

deprecate all use of ToDo arrays
terry.framework.Controller: rename function: getFormattedToDoInfoList() => getToDoList()

terryasdf committed:
Merge pull request #1 from terryasdf/branch-Level-9

Implement command: find

terryasdf committed:
Add and modify Javadoc

remove all CSV related methods

terryasdf committed:
Merge pull request #2 from terryasdf/branch-A-JavaDoc

Add and modify Javadoc

terryasdf committed:
Fix redundant CSV issue

terryasdf committed:
docs\README.md: Add user guide

okkhoy commented on src/main/java/terry/cmd/Cmd.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Parses command lines. Contains a {@link CmdKeyword} and a list of {@link CmdOptArg}.
+ */
+public class Cmd {
```
better to use full form of the work for class names (much clearer); so... `Cmd` -> `Command`

okkhoy commented on src/main/java/terry/entity/ToDo.java:
```diff
@@ -0,0 +1,84 @@
+     * @exception MissingOptArgException when parameter count is less than
+     * {@code leastCount}
+     * */
+    protected static void checkArgCount(List<CmdOptArg> optArgList, int leastCount) throws MissingOptArgException {
```
while not particularly long, this line is probably longer than most others in the codebase.

okkhoy commented on src/main/java/terry/framework/Framework.java:
```diff
@@ -0,0 +1,44 @@
+import terry.cmd.Cmd;
+import terry.cmd.CmdKeyword;
+import terry.cmd.CmdParser;
+import terry.exception.*;
```
Don't use wildcard imports; import the classes as necessary individually

okkhoy commented on src/main/java/terry/framework/Framework.java:
```diff
@@ -0,0 +1,44 @@
+            while (true) {
+                String cmdInput = in.nextLine().strip();
+                // Ignore empty input
+                if (cmdInput.length() == 0) continue;
```
even single line conditional blocks should be enclosed in braces

okkhoy reviewed (commented):
@terryasdf 
Overall, very good code! kudos.
Strongly consider using full form of words for class names (e.g., replace `Msg`, `Cmd` with `Message`, `Command`, etc.,)
I could find a couple of coding standard violations, please fix them before the final submission

terryasdf committed:
Refactor code: replace abbreviations with full words for class and method names

terryasdf committed:
Refactor code: reformat lines

remove all wildcard imports

terryasdf committed:
terry.framework.CommandRunner: Flip the return value for "bye" and other commands

terry.framework.Framework: refactor code

terryasdf committed:
Add a static method to terry.command.CommandOptionArgument: assertLeastOptionCount()

replace argument count checkers with assertLeastOptionCount()

terryasdf committed:
Refactor code

terry.message.Message: Use record class implementation

terryasdf committed:
Merge pull request #3 from terryasdf/branch-Refactor

Branch refactor

terryasdf committed:
Update manifest version: 1.0

terryasdf committed:
Add gradle

terryasdf committed:
Remove irrelevant files

terryasdf committed:
Merge pull request #4 from terryasdf/branch-Gradle

Branch gradle
